skeleton/public/index.html

```diff
@@ -9,8 +9,26 @@
 <body>
   <script src="app.js"></script>
   <script>
-    const app = Elm.Main.init({flags: {}});
-    // register your ports here
+    // The localStorage key to use to store serialized session data
+    const storeKey = "store";
+    
+    const app = Elm.Main.init({
+      flags: {
+        clientUrl: location.origin + location.pathname,
+        store: localStorage[storeKey]
+      }
+    });
+
+    app.ports.saveStore.subscribe((store) => {
+      localStorage[storeKey] = JSON.stringify(store);
+    });
+
+    // Ensure session is refreshed when it changes in another tab/window
+    window.addEventListener("storage", (event) => {
+      if (event.storageArea === localStorage && event.key === storeKey) {
+        app.ports.storeChanged.send(event.newValue);
```

Comment:
Why not use JSON.parse here since you make the stringify on this side as well? You can then use `decodeValue` on a `Decode.Value` to decode on the Elm side rather than `decodeString`

Reply:
Grrr, my comment was mangled. I was saying, I don't want to rely on JSON.parse and JSON.stringify js side because that would mean handling a try/catch for the error case. I've updated the code to always send raw json strings to Elm and proceed with parsing it there.